File: java-advent-of-code/src/main/java/advent/twenty_sixteen/Day08TwoFactorAuthentication.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.utilities.ArrayUtilities;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.List;

public class Day08TwoFactorAuthentication implements DailyProblem<Integer, String> {
    private final int part1Answer;
    private final String part2Answer;

    public Day08TwoFactorAuthentication(String filename, int height, int width) {
        List<String> lines = FileUtilities.readLines(filename, Parsers::TO_STRING);

        char[][] screen = new char[height][];
        for (int row = 0; row != screen.length; row++) {
            screen[row] = new char[width];
            for (int col = 0; col != screen[row].length; col++) {
                screen[row][col] = '.';
            }
        }

        for (String line : lines) {
            if (line.startsWith("rect ")) {
                int indexX = line.indexOf("x");
                int indexS = line.indexOf(" ");
                int w = Integer.parseInt(line.substring(indexS, indexX).trim());
                int h = Integer.parseInt(line.substring(indexX + 1).trim());
                for (int r = 0; r != h; r++) {
                    for (int c = 0; c != w; c++) {
                        screen[r][c] = '#';
                    }
                }
            }
            if (line.startsWith("rotate row y=")) {
                line = line.substring(13);
                int indexOf = line.indexOf("by");
                int row = Integer.parseInt(line.substring(0, indexOf).trim());
                int count = Integer.parseInt(line.substring(indexOf + 2).trim());
                for (int i = 0; i != count; i++) {
                    rotateRow(row, screen);
                }
            }
            if (line.startsWith("rotate column x=")) {
                line = line.substring(16);
                int indexOf = line.indexOf("by");
                int col = Integer.parseInt(line.substring(0, indexOf).trim());
                int count = Integer.parseInt(line.substring(indexOf + 2).trim());
                for (int i = 0; i != count; i++) {
                    rotateCol(col, screen);
                }
            }
        }

        this.part1Answer = ArrayUtilities.count('#', screen);
        this.part2Answer = ArrayUtilities.dumpToString(screen);
    }

    private void rotateCol(int col, char[][] screen) {
        int height = screen.length;

        char c = screen[height - 1][col];
        for (int i = height - 1; i != 0; i--) {
            screen[i][col] = screen[i - 1][col];
        }
        screen[0][col] = c;
    }

    private void rotateRow(int row, char[][] screen) {
        int width = screen[row].length;

        char c = screen[row][width - 1];
        for (int i = width - 1; i != 0; i--) {
            screen[row][i] = screen[row][i - 1];
        }
        screen[row][0] = c;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public String getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day08TwoFactorAuthentication.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

abstract class TwoFactorOperation {
  def update(data: Array[Array[Boolean]]): Array[Array[Boolean]] = ???
}

case class RectOperation(width: Int, height: Int) extends TwoFactorOperation {
  override def update(data: Array[Array[Boolean]]): Array[Array[Boolean]] = {
    for (x <- 0 to width-1) {
      for (y <- 0 to height-1) {
        data(y)(x) = true
      }
    }
    data
  }
}

case class RotateRowOperation(row: Int, count: Int) extends TwoFactorOperation {
  override def update(data: Array[Array[Boolean]]): Array[Array[Boolean]] = {
    val rightIndex = data(0).size -1
    (1 to count).foreach(_ => {
      val temp = data(row)(rightIndex)
      for (col <- rightIndex to 1 by -1) {
        data(row)(col) = data(row)(col-1)
      }
      data(row)(0) = temp
    })
    data
  }
}

case class RotateColumnOperation(col: Int, count: Int) extends TwoFactorOperation {
  override def update(data: Array[Array[Boolean]]): Array[Array[Boolean]] = {
    val bottomIndex = data.size -1
    (1 to count).foreach(c => {
      val temp = data(bottomIndex)(col)
      for (row <- bottomIndex to 1 by -1) {
        data(row)(col) = data(row-1)(col)
      }
      data(0)(col) = temp
    })
    data
  }
}

class Day08TwoFactorAuthentication(filename: String, width: Int, height: Int) extends DailyProblem[Int, String] {
  private val rectPattern = "rect ([a-z0-9]+)x([a-z0-9]+)".r
  private val rowPattern = "rotate row y=([a-z0-9]+) by ([a-z0-9]+)".r
  private val colPattern = "rotate column x=([a-z0-9]+) by ([a-z0-9]+)".r

  private def parser(line: String): TwoFactorOperation =
    line match {
      case rectPattern(width, height) => RectOperation(width.toInt, height.toInt)
      case rowPattern(row, count) => RotateRowOperation(row.toInt, count.toInt)
      case colPattern(col, count) => RotateColumnOperation(col.toInt, count.toInt)
      case _ => throw new IllegalStateException(s"Cannot match '$line'")
    }

  private val commands = FileUtilities.readFile(filename, parser)

  private val result = commands.foldLeft(Array.ofDim[Boolean](height, width))((array, command) => {
    command.update(array)
  })

  override val part1Answer: Int = result.foldLeft(0)((acc, row) => acc + row.count(p => p))
  override val part2Answer: String = result.foldLeft("")((acc, row) => acc + row.foldLeft("")((acc, cell) => acc + (if (cell) "#" else ".")) + "\n")
}
